client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java

```diff
@@ -216,6 +246,18 @@ private void genServerToBlocks(ShuffleBlockInfo sbi, List<ShuffleServerInfo> ser
         partitionToBlocks.put(partitionId, Lists.newArrayList());
       }
       partitionToBlocks.get(partitionId).add(sbi);
+      if (excludeServers != null) {
+        excludeServers.add(ssi);
+      }
+      assignedNum++;
+      if (assignedNum >= replicaNum) {
+        break;
+      }
+    }
+
+    if (assignedNum < replicaNum && withBlocklist) {
+      genServerToBlocks(sbi, serverList, replicaNum - assignedNum,
```

Comment:
Do this cause that one shuffle server will allocate twice?

Reply:
No. Assigned server will be added to `excludeServers`.